Modules/Module_Processing/Module_ASL/xASL_wrp_ProcessM0.m

```diff
@@ -120,10 +119,36 @@ function xASL_wrp_ProcessM0(x)
 
 			% Find the smallest of the previously selected positive nonzero TEs
 			minTE = min(uniqueNonzeroTE);
-			fprintf('Multi-TE M0 present. Using the shortest TE=%.1f ms for M0 calibration.\n', minTE*1000);
+			% JSON values were converted from BIDS (seconds) to Legacy (millisecond) during JSON reading
+			fprintf('Multi-TE M0 present. Using the shortest TE = %.1f ms for M0 calibration.\n', minTE);
+
+			% Find the correct indices and update the jsonM0 parameter
+			indMinTE = jsonM0.Q.EchoTime == minTE;
+			jsonM0.Q.EchoTime = jsonM0.Q.EchoTime(indMinTE);
+			if isfield(jsonM0, 'RepetitionTime') && length(jsonM0.RepetitionTime) > 1
+				jsonM0.RepetitionTime = jsonM0.RepetitionTime(indMinTE);
+			end
+
+			% Select the minimal TEs
+			imM0 = imM0(:, :, :, indMinTE);
+		end
+	end
 
-			% Average M0 image across all these minimal TEs
-			imM0 = xASL_stat_MeanNan(imM0(:,:,:,jsonM0.Q.EchoTime == minTE), 4);
+	% Check if there are multiple different TR times and select the optimal one for M0 quantification
+	if isfield(jsonM0, 'RepetitionTime')
+		uniqueNonzeroTR = unique(jsonM0.RepetitionTime(jsonM0.RepetitionTime > 0));
+		if length(uniqueNonzeroTR) > 1
+			% We currently do not do inversion fitting in multi-TE M0, so we issue a warning and use the TR value closest to 2000ms

```

Comment:
Same comment above for multi-TR M0. We don't do any fitting here.
And we don't issue a warning here.
I would issue a single fprintf('Warning: ...) see also comment above.

Reply:
OK will do...

Reply:
COrrected multi-TE to multi-TR. And print the message similar as for the multi-TE case, but with a warning.